src/Windows/Avalonia.Win32/Interop/Automation/IExpandCollapseProvider.cs

```diff
@@ -9,8 +10,75 @@ namespace Avalonia.Win32.Interop.Automation
     [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
     public interface IExpandCollapseProvider
     {
+#if NET6_0_OR_GREATER
+        public static readonly Guid IID = new("d847d3a5-cab0-4a98-8c32-ecb45c59ad24");
+        public const int VtblSize = 3 + 3;
+#endif
         void Expand();
         void Collapse();
         ExpandCollapseState ExpandCollapseState { get; }
     }
+
+#if NET6_0_OR_GREATER
+    internal static unsafe class IExpandCollapseProviderManagedWrapper
```

Comment:
Why do we need code-generated marshallers checked into the repository?

Reply:
It's not generated code, those wrappers are written by hand :)
Currently the com generator cannot handle properties in interfaces and variant marshalling.

Reply:
We should be using microcom codegen then. The main reason why it wasn't done already is that VARIANT/SAFEARRAY need to be marshalled, but since you are planning to handle them by hand it would be better to just declare those structs in microcom IDL.

Reply:
Sounds fair. Maybe we can just extend the microcom to support VARIANT/SAFEARRAY marshalling?
I'm wondering are there any other pieces of microcom missing now.

Reply:
I think it would be mostly runtime library support to convert between .NET and native representations, not many changes on codegen side are required

Reply:
Seems that MicroCom doesn't recognize array syntax? There're several APIs in a11y using SAFEARRAY.

Reply:
Can it be replaced with IntPtr in the idl file? And then copy-paste SAFEARRAY creation logic from this PR as it is here, only passing/reading pointer to the SAFEARRAY struct.

Reply:
It seems that `Marshal.GetNativeVariantForObject` and  `Marshal.GetObjectForNativeVariant` do work with NativeAOT. So I think that I could add VARIANT support for MicroCom using those.

Do we have unwrapped SAFEARRAY usages anywhere? I don't see relevant `object[]` usages in Avalonia.Win32.

Reply:
Should we treat `VARIANT` arguments as `ref object` ?

Reply:
> Do we have unwrapped SAFEARRAY usages anywhere?

We only use it for marshaling/unmarshaling when crossing the native-managed boundary.

Reply:
I mean, is it used directly anywhere or is it always wrapped into a VARIANT?

Reply:
ahh yeah it's being used directly. for example, `ISelectionProvider.GetSelection` and `IRawElementProviderFragment.GetRuntimeId` which both return a SAFEARRAY directly.

Reply:
> It seems that `Marshal.GetNativeVariantForObject` and `Marshal.GetObjectForNativeVariant` do work with NativeAOT. 

`Marshal.GetNativeVariantForObject` and `Marshal.GetObjectForNativeVariant` require built-in COM which is disabled after NativeAOT and trimming.

Reply:
This snippet does run when compiled with NativeAOT. 

```cs
using System.Runtime.InteropServices;

public static unsafe class Program
{
    public static void Main()
    {

        byte* mem = stackalloc byte[256];
        
        Marshal.GetNativeVariantForObject((int)5, (IntPtr)mem);

        var obj = Marshal.GetObjectForNativeVariant((IntPtr)mem);
        
        Console.WriteLine(obj);

    }
}
```

SAFEARRAY would require manual handling, I guess.